common/filesystem.h

```
@@ -0,0 +1,1374 @@
+  int64_t byte_offset = lseek(fd_, pos, SEEK_SET);
+  if (byte_offset == -1) {
+    return FdError(errno, "File::SeekTo on '{0}'", fd_);
+  }
+  return byte_offset;
+}
+
+inline auto Internal::FileRefBase::SeekFromEnd(int64_t delta)
+    -> ErrorOr<int64_t, FdError> {
+  int64_t byte_offset = lseek(fd_, delta, SEEK_END);
+  if (byte_offset == -1) {
+    return FdError(errno, "File::SeekFromEnd on '{0}'", fd_);
+  }
+  return byte_offset;
+}
+
+inline auto Internal::FileRefBase::ReadToBuffer(
+    llvm::MutableArrayRef<std::byte> buffer)
+    -> ErrorOr<llvm::MutableArrayRef<std::byte>, FdError> {
+  for (;;) {
+    ssize_t read_bytes = read(fd_, buffer.data(), buffer.size());
+    if (read_bytes == -1) {
+      if (errno == EINTR) {
+        continue;
+      }
+      return FdError(errno, "File::Read on '{0}'", fd_);
+    }
+    return buffer.slice(0, read_bytes);
+  }
+}
+
+inline auto Internal::FileRefBase::WriteFromBuffer(
+    llvm::ArrayRef<std::byte> buffer)
+    -> ErrorOr<llvm::ArrayRef<std::byte>, FdError> {
+  for (;;) {
+    ssize_t written_bytes = write(fd_, buffer.data(), buffer.size());
+    if (written_bytes == -1) {
+      if (errno == EINTR) {
+        continue;
+      }
+      return FdError(errno, "File::Write on '{0}'", fd_);
+    }
+    return buffer.drop_front(written_bytes);
+  }
+}
+
+inline auto Internal::FileRefBase::Close() && -> ErrorOr<Success, FdError> {
+  // Put the file in a moved-from state immediately as it is invalid to
+  // retry closing or use the file in any way even if the close fails.
+  int fd = fd_;
+  fd_ = -1;
+
+  int result = close(fd);
+  if (result == 0) {
+    return Success();
+  }
+
+  return FdError(errno, "File::Close on '{0}'", fd);
+}
+
+inline auto Internal::FileRefBase::Destroy() -> void {
+  if (fd_ >= 0) {
+    auto result = std::move(*this).Close();
+    // Debug-only error check. For a non-debug build, we can just leak the
+    // file descriptor.
+    CARBON_DCHECK(result.ok(), "{0}", result.error());
```

This can cause flakiness in debug builds, is that a good idea?

---

I'm torn...

On one hand, I agree, and would rather this just a normal check. There certainly isn't a *performance* concern here.

But this also is very unlikely to corrupt memory or otherwise make it difficult for the program to continue executing. Crashing seems unfortunately severe in that case.

Not sure, open to ideas or suggestions.

---

I was thinking that we should not have a CHECK at all. If the user wants to discard errors then that's what we should do. I don't think a flaky-crashy (since it could occur quite transiently in a CI system for example) filesystem abstraction is doing them any favours.

Let's take this convo to the larger thread below. But I think we should do the same thing here that we do on closing a directory, whatever it is.